RIPS/rip-9997.md

```diff
@@ -0,0 +1,198 @@
+separate transaction.
+
+There is a number of behaviours that define transaction-wide effects in Ethereum.
+This list includes, but is not limited to:
+
+* Tracking `accessed_addresses`
+* [EIP-1283](https://eips.ethereum.org/EIPS/eip-1283) Gas metering for SSTORE
+* [EIP-1153](https://eips.ethereum.org/EIPS/eip-1153) Transient storage opcodes
+
+Any such behaviour has separate effects in the "Validation Virtual Transaction" and "Execution Transaction".
+
+Gas refunds are issued at the end of the entire transaction only.
+
+### Unused gas penalty charge
+
+```
+UNUSED_GAS_PENALTY = 10
+```
+
+Transactions of type `BUNDLE_TRANSACTION_SUBTYPE` that reserve a lot of gas for themselves using `validationGasLimit`,
+`paymasterGasLimit` and `callGasLimit` fields but do not use the reserved gas present a challenge for
+block builders. This is especially demanding in case a gas used by a transaction can be significantly different
+based on its position within a block, as such transactions may cause the block builder to iterate its algorithm
+many times until a fully utilized block is discovered.
+
+A penalty of `UNUSED_GAS_PENALTY` percent of the entire unused gas limit is charged from the
+transaction `sender` or `paymaster`.
+
+The total gas limit is calculated as `totalLimit = validationGasLimit + paymasterGasLimit + callGasLimit`.\
+The `totalGasUsed` is calculated as a sum of all gas used during the transaction.\
+The unused gas is calculated as `unusedGas = totalLimit - totalGasUsed`.
+
+The effects of allowing transactions to specify unrestricted gas limits is shown on this diagram:
+![](../assets/rip-7560/unused_gas_attack_overview.png)
+
+## Rationale
+
+### Using no-op transaction instead of adding a field to the block payload
+
+While in this particular case there is no execution frame that needs to happen for
+a `BUNDLE_HEADER_SUBTYPE` transaction, in other similar scenarios it is possible we
+will require some execution for a header transaction.
+
+For example, in (RIP-xxxx: Signature Aggregation)[todo] the header transaction contains
+the execution of the aggregted signature verification code.
+
+For consistency with potential future extensions to the RIP-7560 it seems reasonable to
+introduce a no-op transaction subtype.
+
+### Breaking up a transaction flow into non-atomic pieces
+
+While changing the transaction flow is a departure from the norm for Ethereum transactions,
+in practice this change is both very useful and not very complex.
+
+The validation frames can be easily constrained with both gas limits and opcode bans.
+The entire validation section of a correctly built bundle can therefore be protected from containing
+mutually exclusive transactions.
+The task of building a block becomes efficient and can be parallelized with no risk of DoS attacks.
+
+For the EVM implementation the validation frames behave almost identical to a standalone transaction,
```

Comment:
I think we'll still have to preserve some context between validation and execution.  Warm-status for accounts and slots, and transient storage.  Otherwise AA becomes needlessly more expensive, and transient storage behavior becomes non-intuitive.  We previously calculated that the context that the validator has to keep in memory for such validation will be sufficiently small, because of the high gas costs associated with inflating it.  Does this add unreasonable complexity?

Reply:
It kind of does when I think about it, but not too much.
We would need to pre-apply the warmings and TSTOREs and such from the validation frame to the execution frame.
So that if a contract uses `TSTORE` "slots" in validations of transactions A and B, would we keep two separate transient storage slots so transaction A sees value from A and transaction B sees value from B.
Should I update the RIP?

Reply:
I think we should try to avoid keeping more context between validation and execution, especially not dynamic context.
**global-warming** reduces the need to keep warm state between validation and execution (it is across all validations and executions)
TSTORE becomes problematic if there are several TXs of the same sender.